experimental/examples/imageviewer/shared/src/iosMain/kotlin/example/imageviewer/storage/IosImageStorage.ios.kt

```diff
@@ -0,0 +1,180 @@
+                    }.sortedByDescending {
+                        it.timeStampSeconds
+                    }
+            )
+        } else {
+            fileManager.createDirectoryAtURL(savePictureDir, true, null, null)
+        }
+    }
+
+    private fun makeFileUrl(fileName: String) =
+        savePictureDir.URLByAppendingPathComponent(fileName)!!
+
+    private fun readStringFromFile(fileName: String): String =
+        NSString.stringWithContentsOfURL(
+            url = makeFileUrl(fileName),
+            encoding = NSUTF8StringEncoding,
+            error = null,
+        ) as String
+
+    private fun String.writeToFile(fileName: String) =
+        writeToURL(makeFileUrl(fileName))
+
+    private fun readPngFromFile(fileName: String) =
+        NSData.dataWithContentsOfURL(makeFileUrl(fileName))
+
+    private fun NSData.writeToFile(fileName: String) =
+        writeToURL(makeFileUrl(fileName), true)
+
+    override fun saveImage(picture: PictureData.Camera, image: PlatformStorableImage) {
+        ioScope.launch {
+            UIImageJPEGRepresentation(image.rawValue.resizeToThumbnail(), 0.6)
+                ?.writeToFile(picture.thumbnailPngFile)
+            pictures.add(0, picture)
+            UIImageJPEGRepresentation(image.rawValue.resizeToBig(), 0.6)
+                ?.writeToFile(picture.pngFile)
+            val jsonStr = Json.Default.encodeToString(picture)
+            jsonStr.writeToFile(picture.jsonFile)
+        }
+    }
+
+    override suspend fun getThumbnail(picture: PictureData.Camera): ImageBitmap =
+        ioScope.async {
+            val jpgRepresentation = readPngFromFile(picture.thumbnailPngFile)!!
+            val byteArray: ByteArray = ByteArray(jpgRepresentation.length.toInt()).apply {
+                usePinned {
+                    memcpy(it.addressOf(0), jpgRepresentation.bytes, jpgRepresentation.length)
+                }
+            }
+            Image.makeFromEncoded(byteArray).toComposeImageBitmap()
+        }.await()
+
+    override suspend fun getImage(picture: PictureData.Camera): ImageBitmap =
+        ioScope.async {
+            fun getFileContent() = readPngFromFile(picture.pngFile)
+            var jpgRepresentation: NSData? = getFileContent()
+            while (jpgRepresentation == null) {
+                yield()
+                jpgRepresentation = getFileContent()
+            }
+            val byteArray: ByteArray = ByteArray(jpgRepresentation.length.toInt()).apply {
+                usePinned {
+                    memcpy(it.addressOf(0), jpgRepresentation.bytes, jpgRepresentation.length)
+                }
+            }
+            Image.makeFromEncoded(byteArray).toComposeImageBitmap()
+        }.await()
+
+}
+
+private fun UIImage.resizeToThumbnail(): UIImage {
+    val targetScale = maxOf(
+        storableThumbnailSize.toFloat() / size.useContents { width },
+        storableThumbnailSize.toFloat() / size.useContents { height },
+    )
```

Comment:
It's a good candidate for helper function extraction

Reply:
Yeah, a good suggestion, but also it is better to postpone it